### DISCUSSION TITLE
tutorials/lift-off-part3/05-resolver-chains

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#26] [course feedback]
# tutorials/lift-off-part3/05-resolver-chains

https://www.apollographql.com/tutorials/lift-off-part3/05-resolver-chains

**Comment by essaji** [comment]
Aren't we creating N times call problem when fetching modules in a chain?

**Reply by mabuyo** [reply]
Hey @essaji! Can you clarify what you mean by fetching modules in a chain? 

The [REST endpoint for `/track/:id/modules`](https://odyssey-lift-off-rest-api.herokuapp.com/docs/#/Tracks/get_track__id__modules) returns all of the module details we need for a given track, so we're not making one request for each module, we just do the one call for a given track. Is that what you're referring to?

**Reply by zakiRefai** [reply]
I think what they are asking is technically true. If you were to call a track and want to get the details of a module as well, you would be doing it N + 1. However, the upside to this approach is that everything is separated. We can get a track without module details, and not call the endpoint for module details. If we want module details, those calls are made but the data is stored for faster retrieval the next time around. Is this correct @maybuyo?

**Reply by mabuyo** [reply]
I think we might be talking about a few different examples, so hopefully my explanations below will provide some clarity!

**Example 1:** Query for one track and the name of all of its modules

```graphql
query track(id: "c_0") {
  title
  modules {
    title
  }
}
```

With the way the resolvers are set up (ie, we have a separate resolver for `Track.modules`), this would only ever result in 2 calls. One for `GET /track/:id` and one for `GET /track/:id/modules`.

If we were to put all of the fetching logic in one resolver (`Query.track`), we would still only have 2 calls, but the downside is that if we had a query that did _not_ ask for modules information, the call to `GET /track/:id/modules` would still be made. (If we had a separate resolver, we would not have this problem!)

**Example 2:** Query for a _list of tracks_. For each track, get the name of all its modules.

```graphql
query tracksForHome {
  title
  modules {
    title
  }
}
```

This one is a different example from the one above. (Also note that we don't expect the client app to make this query because it's not part of any of the UI designs!)

If we have a separate resolver for `Track.modules`, we would have N + 1 calls. 1 to get the list of tracks `GET /tracks` and then N (for however many tracks there are) to call `GET /track/:id/modules`.

If we were to put all of the fetching logic in one resolver (`Query.tracksForHome`), we would still also have N + 1 calls, and we face the same downside as in example 1.

**Takeaway**: Writing separate resolvers is the way to go! 

Lastly, for your point:

> but the data is stored for faster retrieval the next time around

Yes, that's the benefit of using the `RESTDataSource` class! We talk about it in [this lesson in Lift-off II](https://www.apollographql.com/tutorials/lift-off-part2/03-apollo-restdatasource). You can also read more about the `RESTDataSource` caching [in the Apollo documentation](https://www.apollographql.com/docs/apollo-server/data/fetching-rest/#caching).

Please let me know if we can clarify anything else! We can add this topic on our list to improve the course for future iterations if it was helpful :)

**Comment by EvgheniAnachii** [comment]
Hi,
In this lesson we had to add in our resolver:
```modules: ({id}, _, {dataSources}) => {
  return dataSources.trackAPI.getTrackModules(id);
},```
in our schema we defined `modules` as the following: `modules: [Module!]!`. I use TS, so even after `npm run generate` I get TS error in my resolver for `modules` complaining about `id`. If we take `id` from parent, why TS is complaining? Did we miss to indicate that there is an `id` on parent?

**Reply by EvgheniAnachii** [reply]
Sorry for the formatting.
In this lesson we had to add in our resolver:

```
modules: ({id}, _, {dataSources}) => {
  return dataSources.trackAPI.getTrackModules(id);
},
```

in our schema we defined `modules` as the following: `modules: [Module!]!`. I use TS, so even after `npm run generate` I get TS error in my resolver for `modules` complaining about `id`. If we take `id` from parent, why TS is complaining?
So, TS doesn't like destracturing of the `id` from "parent" argument:
```
modules: ({id}, _, {dataSources}) => {
    return dataSources.trackAPI.getTrackModules(id);
},
```

**Reply by EvgheniAnachii** [reply]
Ok, I've sorted it out. My `modules` resolver was under `Query` not under `Track`.

**Reply by mabuyo** [reply]
Happy to hear you've found the issue @EvgheniAnachii !

**Comment by aahadaazar** [comment]
Hello

isnt this line:
const track = dataSources.trackAPI.getTrack(id);
supposed to have an await?

**Reply by mabuyo** [reply]
Hey @aahadaazar - good catch! This is just an example however, to show what we _could_ do. In the end, we will create a new resolver and the `track` resolver function stays the same (immediately returning the results of that call).

**Comment by ohmprakash1506** [comment]
i am getting the following error in my code 

ERROR] 16:35:36 ⨯ Unable to compile TypeScript:
src/resolvers.ts(22,5): error TS2322: Type '{ author: ({ authorId }: TrackModel, _: {}, { dataSources }: DataSourceContext) => Promise<AuthorModel>; modules: ({ id }: { id: any; }, _: any, { datasource }: { datasource: any; }) => any; }' is not assignable to type 'TrackResolvers<DataSourceContext, TrackModel>'.
  Object literal may only specify known properties, and 'modules' does not exist in type 'TrackResolvers<DataSourceContext, TrackModel>'.

the code is as follows :

import { Resolvers } from "./types";

export const resolvers: Resolvers = {
  Query: {
    // returns an array of Tracks that will be used to populate the homepage grid of our web client
    tracksForHome: (_, __, { dataSources }) => {
      return dataSources.trackAPI.getTracksForHome();
    },

    // get a single track by ID, for the track page
    track: async (_, { id }, { dataSources }) => {
      const track = dataSources.trackAPI.getTrack(id);
      const modules = await dataSources.trackAPI.getTrackModules(id);

      return { ...track, modules };
    },
  },
  Track: {
    author: ({ authorId }, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor(authorId);
    },
    modules: ({ id }, _, { datasource }) => {
      return datasource.trackAPI.getTrackModules(id);
    },
  },
};

@ mabuyo can you help me out fix this ?

**Reply by mabuyo** [reply]
Hey there @ohmprakash1506. The error is letting you know that the types aren't matching in a specific line. Which line is it pointing to? It's difficult to tell in the code snippet you posted.

That said, I can also see that you've got a typo in the resolver. It shoudl be `dataSources` (with an s at the end). Hope that helps!

```ts
modules: ({id}, _, {dataSources}) => {
  return dataSources.trackAPI.getTrackModules(id);
},
```

**Reply by emilioramirezeguia** [reply]
I had the same error. The problem was that we were missing square brackets [] following the "ModuleModel" type in the `getTrackModules` method.

```
getTrackModules(trackId: string) {
    return this.get<ModuleModel[]>(`track/${trackId}/modules`);
}
```

**Comment by emilioramirezeguia** [comment]
@mabuyo I'm a little confused on how resolvers determine who the `parent` is. In the explanation it says that it refers to the returned data of the preceding resolver function in the chain. If we added the Track.modules right below Track.author, wouldn't the parent, and therefore the returned data, be _author_?

**Reply by lizhennessy** [reply]
Hey there @emilioramirezeguia ! Resolver chains are probably _the trickiest_ part of GraphQL; it can be hard to visualize exactly where each piece is coming from.

Let's say we had the following query:

```
query GetTrackAuthorAndModules {
  track(id: $trackId) {
    id
    title
    author {
      id
      name
    }
    modules {
      id
      title
      length
    }
  }
}
```

First off, we're calling `Query.track`. This returns a `Track` type, for which we're asking for a number of fields: `Track.id`, `Track.title`, `Track.author`, `Track.modules`. As we saw in [lesson 4](https://www.apollographql.com/tutorials/lift-off-part3/04-resolver-args-parameter), calling the `Query.track` resolver immediately makes a request to our REST API, giving a nice JSON object we can pluck properties like `id` and `title` from — kind of like the example below!

```
{
  "id": "c_0",
  "title": "Cat-stronomy, an introduction",
  "authorId": "cat-1",
  "modules": [
    "l_0",
    "l_1",
    "l_2",
  ]
}
```

When we get to the point of resolving `Track.author`, we already have the response from our REST API for the particular track we're querying. This object is what arrives in our `Track.author` resolver as the `parent` argument: it's the blob of data representing the track, like the one shown above. It's how we can get access to the `authorId` that came back in the response.

The same is true when we get to `Track.modules`.  (Notice that the `author` and `modules` fields in the query are on the same "level" - they both belong to the `Track` type, so `modules` isn't nested _under_ the `author` subfields). The resolver for `Track.modules` will receive the same JSON object from the REST API as the `Track.author` did: so we're able to traverse that list of module IDs and make any other requests necessary.

I hope that this has helped! Please reach out if you have any other questions. :)

**Reply by emilioramirezeguia** [reply]
Hey @lizhennessy! Thanks for your reply. That makes perfect sense. I was confused in thinking the parent-child relationship was determined by the literal placement of the resolver functions in the `resolvers.ts` file, but it really depends on the order in which you query these fields in the client and the relationship between them in the schema definition, right?

**Reply by lizhennessy** [reply]
Exactly right, @emilioramirezeguia ! Before either of the `Track.modules` or `Track.author` fields _can_ be resolved, `Query.track` has to be resolved first! So that builds our chain. :)

**Reply by emilioramirezeguia** [reply]
Awesome! It's clear now. Thank you for your help 🙂

**Comment by LingVuDev** [comment]
In this track resolver example:
```
const track = dataSources.trackAPI.getTrack(id);
```
The `await` keyword is missing when calling the `trackAPI` function.

**Reply by mabuyo** [reply]
Ahh, that code snippet is an example not meant to be copied, but thank you for catching that! Will update :)

**Comment by arvindsid98** [comment]
query GetTrackAuthorAndModules {
  track(id: $trackId) {
    id
    title
    author {
      id
      name
    }
    modules {
      id
      title
      length
    }
  }
}

Is it possible to access author.name inside modules?

**Reply by mabuyo** [reply]
Hi @arvindsid98 ! Do you mean a query like this?

```graphql
query GetTrackAuthorAndModules {
  track(id: $trackId) {
    id
    title
    author {
      id
      name
    }
    modules {
      id
      title
      length
      author {
        name
      }
    }
  }
}
```

That will only be possible if we update the schema and code. The `Module` type would need an `author` field returning type `Author`. The resolver function for this field would need to do an additional call to the data source, since the module object doesn't have an `authorId` property (unlike the track object).

Hope that helps!

**Comment by Zain-ul-din** [comment]
Our restful-APIS are returning an array of modules id 

```
"modules": [
    "l_10",
    "l_11",
    "l_12",
    "l_13",
    "l_14"
  ]
```

How does resolver knows it to convert into `{ id: "l_10}` ...

**Reply by Zain-ul-din** [reply]
My bad, i just studied this `id` is coming from track and it is `track.id`

**Comment by SergiiBurgazlieiev** [comment]
I think we forgot to edit ```server/codegen.ts``` file in this part of the tutorial and add it to the mapper object to replace a GraphQL type usage with a custom type ModuleModel: 
```
config: {
  mappers: {
     ...
     Module: './models#ModuleModel'
  }
}
```

**Comment by manovinayak** [comment]
Hi,
This is Mano. 

Theses tutorials were really great. I have already implemented resolver chains for queries at work.

My question is: Can Mutations can be chained in the same pattern? And is it a standard approved pattern in graphql.


Would like to know if this resolver chain can be applied to Mutations as well. 

Let's suppose, I would want to create a Track and then update the Author of the track.

Resolvers:
CreateTrack
UpdateTrack

Is it possible to chain the resolvers, so that they can executed in sequence?

The reason I am asking is because, I do not see much of tutorials on Mutation chaining and most or all of them talk only about Query Resolver chaining.

Regards,
Mano

**Reply by lizhennessy** [reply]
Hi there @manovinayak ! This is a great and interesting question. Whenever I use resolver chains, it's because I need some extra data to be fetched as part of the response; so to use the typical example of the query, we could query a whole bunch of tracks from a single endpoint, but if our query includes the `author` field, we might have a different resolver that's responsible for reaching out to a different endpoint for supplemental data.

In the case of mutations, I think it's important to remain just as focused and atomic. For instance if you had one `Mutation` field called `createTrack`, and another called `updateTrack`, your frontend consumers would likely expect each of these fields to perform just a single operation. In this way, the contract remains very clear between frontend and back; there aren't any (we hope) hidden side effects that will impact data down the line. So with this example, I think I'd lean toward putting the sequential logic in the service layer. Perhaps our goal is to create a new track with user-supplied data. If we then wanted to update the newly created track with supplemental data, such as we might require for storing in the database, I'd most likely keep those transformations as close to the database as possible. 

We can still of course use resolver chains with mutations, for example if our `createTrack` field returns a `Track`. If we write a query like the following:

```
mutation CreateTrack($trackInput: TrackInput) {
    success
    track {
        id
        name
        author {
            name
            allTracksProduced {
                      name
            }
      }
}
```

Then perhaps our initial mutation field resolver is responsible for returning _just_ the `success` and track `id` and `name`; but our mutation might include additional fields about the Track that would then need to be provided by `Track` specific resolvers, such as `author`. Furthermore, we might have nested fields like `allTracksProduced` which we could use to see what else that particular author had produced.

So in this sense, we can certainly make use of resolver chains with mutations, and delegate specific logic to different functions. I wouldn't necessarily use them to trigger additional backend changes that aren't immediately apparent to the consumer, however. I hope this explanation helped a little bit! It was a great question. Please reach out if there's anything else we can help with!

**Reply by manovinayak** [reply]
Hi @lizhennessy ,

Thank you so much for the detailed explanation 😄. It's very clear 👍.

The main reason for me to think about resolver chains was to reduce the number of network calls from the client side. 
I have achieved that for queries using resolver chains. But for mutations, looks like we need to sequence the backend API calls from GraphQL BFF and return the final response to the client after the final successful backend API/Database call.

May aim is to reduce the number of network calls and especially in mobile networks as this makes a big difference.

On the other hand, Mutation followed by Query(s) can be chained, as you have explained.

2 years ago, these tutorials helped me better understand GraphQL more than any other tutorial. Here I am back the place where I started to ask a question again on the topic that I have been ruminating for a while 😄 .

Thank you so much. Apollo GraphQL is a great platform 🙇‍♂️ 


Thanks and Regards,
Mano

**Reply by lizhennessy** [reply]
@manovinayak that's great additional context. I think it definitely depends on the goal of each mutation field: adding a new track, versus updating a track, for example. The mutations give us a good way for clients using our app on the frontend to control or at the very least trigger changes on the backend - but how we get there, on the server side at least, can definitely vary! (For instance we might have a controller layer, a service layer, and so on, where additional ripple effects might occur, all the while originating from a single user-triggered action!) I'm so glad that you've had such a good experience on Odyssey. We love having learners like you and of course hearing your feedback so that we can improve our content and experience for everyone. :)

**Comment by TomasB** [comment]
I which you guys really would have stuck to those boring library / book / author instead of module / track / etc.. Just looking at this without any prior context, I have no idea what is track and what is module and what are the relations between them, thus instead of using brainpower to consume the material, the attention is drawn elsewhere.

**Reply by TomasB** [reply]
*wish (no edit in comments?)

**Reply by lizhennessy** [reply]
Hi there @TomasB ! Thanks for the feedback. This is a great point, and in other courses we've definitely stuck to domains like `Book`/`Author` that have some distance from the topic of the course. I'll bring this feedback back to the team for review!

**Comment by samuelpucat** [comment]
Hello there,
hypothetically speaking, if there wasn't `GET:/track/{id}/modules` API, but only `GET:/module/{id}` API.
Should I pass `modules: string[]` from `Track.modules` resolver to `trackAPI.getTrackModules` and there call  `GET:/module/{id}` N times?
Or is there a better way to do this with such constraint?

**Comment by AtGbemiga** [comment]
Why do this which is correct, 
```ts
Track: {
    author: ({ authorId }, _, { dataSources }) => {
      // console.log(contextValue);

      // console.log(parent); // won't work cos destructured
      return dataSources.trackAPI.getAuthor(authorId);
    },

    modules: ({id}, _, {dataSources}) => {
      return dataSources.trackAPI.getTrackModules(id);
    },
  },
```
instead of this which is wrong
```ts
Query: {
    // returns an array of Tracks that will be used to populate
    // the homepage grid of our web client
    tracksForHome: (_, __, { dataSources }) => {
      // get all tracks, will be used to populate the homepage grid of our web client

      // trackAPI is not custom, it is made by me
      return dataSources.trackAPI.getTracksForHome();
    },

    // Will get error here w/o first running the codegen command
    track: (parent, { id }, { dataSources }, info) => {
      return dataSources.trackAPI.getTrack(id);
    },

    modules: ({id}, _, {dataSources}) => {
      return dataSources.trackAPI.getTrackModules(id);
    },
  },
```

P.S How do I know if to place the resolver in the query or a custom resolver type?

**Reply by mabuyo** [reply]
Hi there @AtGbemiga ! I don't quite understand what you're trying to compare with the code snippets you've provided. Can you clarify your question?

**Reply by AtGbemiga** [reply]
Hi. I want to understand how to determine whether a resolver should go inside Query or inside a custom resolver type like Track.

**Reply by mabuyo** [reply]
@AtGbemiga thanks for clarifying! **The resolvers map needs to follow the shape of the schema.** 

In your second code snippet, I see you have `modules` under `Query`, but since the schema is shaped like this:

```graphql
type Track {
# ... other Track fields
  modules: [Module!]!
}
```

That second code snippet's resolver map wouldn't match the shape of the schema. I hope that makes sense! Let me know if you have any follow-up questions.

**Reply by AtGbemiga** [reply]
Yes, this "The resolvers map needs to follow the shape of the schema" answers my question.
